src/webgl/ShaderGenerator.js

```diff
@@ -1638,3 +1638,480 @@ export default shadergenerator;
+ *
+ * Note: This hook is currently limited to per-vertex operations; storing variables for later use is not supported.
+ *
+ * This hook is available in:
+ * - <a href="#/p5/baseColorShader">baseColorShader()</a>
+ * - <a href="#/p5/baseMaterialShader">baseMaterialShader()</a>
+ * - <a href="#/p5/baseNormalShader">baseNormalShader()</a>
+ * - <a href="#/p5/baseStrokeShader">baseStrokeShader()</a>
+ *
+ * @param {function} callback
+ *        A callback function which is called before each vertex is processed.
+ */
+
+/**
+ * @function afterVertex
+ * @description
+ * Registers a callback to run custom code at the very end of the vertex shader. This hook can be used inside <a href="#/p5/baseColorShader">baseColorShader()</a>.modify() and similar shader modify calls to perform cleanup or final calculations after all vertex processing is done. The callback receives no arguments.
+ *
+ * Note: This hook is currently limited to per-vertex operations; storing variables for later use is not supported.
+ *
+ * This hook is available in:
+ * - <a href="#/p5/baseColorShader">baseColorShader()</a>
+ * - <a href="#/p5/baseMaterialShader">baseMaterialShader()</a>
+ * - <a href="#/p5/baseNormalShader">baseNormalShader()</a>
+ * - <a href="#/p5/baseStrokeShader">baseStrokeShader()</a>
+ *
+ * @param {function} callback
+ *        A callback function which is called after each vertex is processed.
+ */
+
+/**
+ * @function beforeFragment
+ * @description
+ * Registers a callback to run custom code at the very start of the fragment shader. This hook can be used inside <a href="#/p5/baseColorShader">baseColorShader()</a>.modify() and similar shader modify calls to set up variables or perform calculations that affect every pixel before color calculations begin. The callback receives no arguments.
+ *
+ * This hook is available in:
+ * - <a href="#/p5/baseColorShader">baseColorShader()</a>
+ * - <a href="#/p5/baseMaterialShader">baseMaterialShader()</a>
+ * - <a href="#/p5/baseNormalShader">baseNormalShader()</a>
+ * - <a href="#/p5/baseStrokeShader">baseStrokeShader()</a>
+ *
+ * @param {function} callback
+ *        A callback function which is called before each fragment is processed.
+ *
+ * @example
+ * <div modernizr='webgl'>
+ * <code>
+ * let myShader;
+ * function setup() {
+ *   createCanvas(200, 200, WEBGL);
+ *   myShader = baseColorShader().modify(() => {
+ *     beforeFragment(() => {
+ *       // Set a value for use in getFinalColor
+ *       this.brightness = 0.5 + 0.5 * sin(millis() * 0.001);
+ *     });
+ *     getFinalColor(color => {
+ *       // Use the value set in beforeFragment to tint the color
+ *       color[0] *= this.brightness; // Tint red channel
```

Comment:
Looks like we still have some uses of array indexing. This should be `color.r *= ...`

Reply:
(Although maybe don't worry about this particular instance as I mentioned earlier that we probably want to `@private` this particular method.)